ixmp/core.py

```diff
@@ -583,9 +592,20 @@ def timeseries(self, region=None, variable=None, unit=None, year=None,
         df['model'] = self.model
         df['scenario'] = self.scenario
 
+        # drop `subannual` column if not requested (False) or required ('auto')
+        if subannual is not True:
+            has_subannual = any(df['subannual'].unique() != ['Year'])
```

Comment:
nit: not sure but maybe
```
df[df['subannual'] != 'Year'].any()
```
is cheaper? @khaeru do you have an idea?

Reply:
I doubt that the alternative would be much more efficient, so I'd keep the current implementation because it is easier to read:

> any of df-subannual not equals year

compared to

> df of df-subannual not equals year - and now any of that...

Reply:
```python
In [1]: s = pd.Series(list(range(1000)), dtype='str')
In [2]: s[500] = 'Year'
In [3]: df = pd.DataFrame(dict(value=123., subannual=s, unit='kg'))
In [4]: %timeit any(df['subannual'].unique() != ['Year'])
139 µs ± 2.67 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [5]: %timeit df[df['subannual'] != 'Year'].any()
807 µs ± 7.21 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [6]: %timeit (df['subannual'] != 'Year').any()
163 µs ± 6.84 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [7]: %timeit set(df['subannual'].unique()) > {'Year'}
157 µs ± 3.73 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [8]: %timeit any(df['subannual'].unique() != 'Year')
142 µs ± 6.76 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [9]: %timeit (df['subannual'] != 'Year').any()
184 µs ± 10.7 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [10]: %timeit not (df['subannual'] == 'Year').all()
155 µs ± 3.58 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [11]: %timeit ts = df['subannual'].unique(); len(ts) == 1 and ts[0] == 'Year'
111 µs ± 818 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [12]: %timeit not all(df['subannual'] == 'Year')  
116 µs ± 3.39 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

There's not a lot of difference, indeed. For readability, I would slightly prefer the last one, which is almost plain English: "not all subannual values are 'Year'"

Reply:
changed to most-close-to-spoken as suggested - thanks for coming up with a dozen alternatives, very impressive...

any idea why only one approach is so much less efficient?

Reply:
> any idea why only one approach is so much less efficient?

5 starts by computing a pd.Series of True/False: `df['subannual'] != 'Year'`. Then it uses that Series to index the DataFrame: `df[...]`. This operation creates a whole new pd.DataFrame (or view) with only the rows that have True in the first step. It's manipulating _all_ the columns, which I guess is why it's slower.

Reply:
@khaeru thanks a lot for comparison!
You are right `df[df['subannual'] != 'Year'].any()` has a mistake, when suggested I meant `(df['subannual'] != 'Year').any()`.
I thought it could be difference between python function `all`/`any` and methods `all`/`any` of Pandas, but it doesn't seem to be the case.
Here are some results for different dataframe sizes:
![index](https://user-images.githubusercontent.com/79057/77854150-c930ee80-71e8-11ea-9069-f84c816c3eed.png)